themes/survey/vanilla/views/subviews/header/google_analytics.twig

```diff
@@ -15,26 +15,46 @@
     See: survey settings
 #}
 {% if aSurveyInfo.googleanalyticsapikey  is defined and aSurveyInfo.googleanalyticsapikey and aSurveyInfo.googleanalyticsstyle %}
-    {% if aSurveyInfo.googleanalyticsstyle  is defined and aSurveyInfo.googleanalyticsstyle == 1 %}
-        <script>
-        (function(i,s,o,g,r,a,m){ i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
-            (i[r].q=i[r].q||[]).push(arguments) },i[r].l=1*new Date();a=s.createElement(o),
-            m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
-        })(window,document,'script','//www.google-analytics.com/analytics.js','ga');
+    {# Check if the tracking ID is a Universal Analytics (legacy) or a Google Analytics 4 ID #}
+    {% if aSurveyInfo.googleanalyticsapikey starts with 'UA-' %}
+        {# Legacy UA #}
+        {% if aSurveyInfo.googleanalyticsstyle  is defined and aSurveyInfo.googleanalyticsstyle == 1 %}
+            <script>
+            (function(i,s,o,g,r,a,m){ i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
+                (i[r].q=i[r].q||[]).push(arguments) },i[r].l=1*new Date();a=s.createElement(o),
+                m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
+            })(window,document,'script','//www.google-analytics.com/analytics.js','ga');
 
-        ga('create', '{{aSurveyInfo.googleanalyticsapikey}}', 'auto');  // Replace with your property ID.
-        ga('send', 'pageview');
-        </script>
+            ga('create', '{{aSurveyInfo.googleanalyticsapikey}}', 'auto');  // Replace with your property ID.
+            ga('send', 'pageview');
+            </script>
+        {% else %}
+            {% set trackUrl = getGoogleAnalyticsTrackingUrl(aSurveyInfo.sid, aSurveyInfo.trackUrlPageName ?? '') %}
+            <script>
+            (function(i,s,o,g,r,a,m){ i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){ (i[r].q=i[r].q||[]).push(arguments) }
+            ,i[r].l=1*new Date();a=s.createElement(o), m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
+            })(window,document,'script','//www.google-analytics.com/analytics.js','ga');
+
+            ga('create', '{{ aSurveyInfo.googleanalyticsapikey }}', 'auto');
+            ga('send', 'pageview');
+            ga('send', 'pageview', '{{ trackUrl }}');
+            </script>
+        {% endif %}
     {% else %}
-        {% set trackUrl = getGoogleAnalyticsTrackingUrl(aSurveyInfo.sid, aSurveyInfo.trackUrlPageName ?? '') %}
+        {# GA4 #}
+        <!-- Google tag (gtag.js) -->
+        <script async src="https://www.googletagmanager.com/gtag/js?id={{aSurveyInfo.googleanalyticsapikey}}"></script>
```

Comment:
We don't encode googleanalyticsapikey in controller, and don't filter (or fix) in model rules

I 'm sure i can find XSS here.

Reply:
Not same method : now it's on an url, before in a JS var.